linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/LinkageMonitor.java

```diff
@@ -16,22 +16,91 @@
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import java.io.IOException;
+import org.eclipse.aether.RepositoryException;
+import org.eclipse.aether.RepositorySystem;
+import org.eclipse.aether.RepositorySystemSession;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.resolution.VersionRangeRequest;
+import org.eclipse.aether.resolution.VersionRangeResolutionException;
+import org.eclipse.aether.resolution.VersionRangeResult;
+
 /**
  * Linkage Monitor detects new linkage errors caused by locally-installed snapshot artifacts for a
  * BOM (bill-of-materials).
  */
 public class LinkageMonitor {
 
-  public static void main(String[] arguments) {
+  public static void main(String[] arguments) throws RepositoryException, IOException {
     if (arguments.length < 1) {
       System.err.println(
           "Please specify BOM coordinates. Example: com.google.cloud:libraries-bom:1.2.1");
       System.exit(1);
     }
     String bomCoordinates = arguments[0];
-    System.out.println("Linkage Monitor for " + bomCoordinates);
-    // TODO(#681): Run Linkage Checker for the BOM specified in argument
-    // TODO(#682): Copy the BOM with locally-installed snapshot versions
-    // TODO(#683): Display new linkage errors caused by snapshot versions if any
+    Bom baseline = RepositoryUtility.readBom(bomCoordinates);
+    Bom snapshot = copyWithSnapshot(baseline);
+
+    ImmutableSet<SymbolProblem> problemInBaseline =
+        LinkageChecker.create(baseline).findSymbolProblems().keySet();
+    ImmutableSet<SymbolProblem> problemsInSnapshot =
+        LinkageChecker.create(snapshot).findSymbolProblems().keySet();
+
+    if (problemInBaseline.containsAll(problemsInSnapshot)) {
+      // No new symbol problems introduced by snapshot BOM. Returning success.
+      return;
+    } else {
+      // TODO(#683): Display new linkage errors caused by snapshot versions if any
+      System.err.println("There are one or more new new linkage errors in snapshot versions:");
+      System.err.println(Sets.difference(problemInBaseline, problemsInSnapshot));
+      System.exit(1);
+    }
+  }
+
+  /**
+   * Returns a copy of {@code bom} replacing its managed dependencies that have locally-installed
+   * snapshot versions.
+   */
+  private static Bom copyWithSnapshot(Bom bom) throws VersionRangeResolutionException {
+    ImmutableList.Builder<Artifact> managedDependencies = ImmutableList.builder();
+
+    RepositorySystem repositorySystem = RepositoryUtility.newFileRepositorySystem();
+    RepositorySystemSession session = RepositoryUtility.newSession(repositorySystem);
+
+    for (Artifact managedDependency : bom.getManagedDependencies()) {
+
+      String snapshotVersion = findSnapshotVersion(repositorySystem, session, managedDependency);
+      if (snapshotVersion == null) {
+        managedDependencies.add(managedDependency);
+      } else {
+        managedDependency.setVersion(snapshotVersion);
```

Comment:
Don't you still need to add `managedDependency` to `managedDependencies`?

Reply:
Yes, because we want a copy of Bom object

Reply:
Fixed